rust/kernel/hrtimer.rs

```diff
@@ -0,0 +1,279 @@
+//!        type Receiver<'a> = Pin<&'a IntrusiveTimer>;
+//!
+//!        fn run(this: Self::Receiver<'_>) {
+//!            pr_info!("Timer called\n");
+//!            this.flag.store(true, Ordering::Relaxed);
+//!        }
+//!    }
+//!
+//!    impl_has_timer! {
+//!        impl HasTimer<Self> for IntrusiveTimer { self.timer }
+//!    }
+//!
+//!    stack_pin_init!(let timer = IntrusiveTimer::new());
+//!    let flag = timer.flag.clone();
+//!
+//!    timer.into_ref().schedule(200_000_000);
+//!    while !flag.load(Ordering::Relaxed) {}
+//!    pr_info!("Done\n");
+//! ```
+//!
+//! C header: [`include/linux/hrtimer.h`](srctree/include/linux/workqueue.h)
+
+use core::{marker::PhantomData, pin::Pin};
+
+use crate::{init::PinInit, prelude::*, types::Opaque};
+
+/// A timer backed by a C `struct hrtimer`
+///
+/// # Invariants
+///
+/// * `self.timer` is initialized by `bindings::hrtimer_init`.
+///
+#[repr(transparent)]
+#[pin_data(PinnedDrop)]
+pub struct Timer<T> {
+    #[pin]
+    timer: Opaque<bindings::hrtimer>,
+    _t: PhantomData<T>,
+}
+
+// SAFETY: A `Timer` can be moved to other threads and used from there.
+unsafe impl<T> Send for Timer<T> {}
+
+// SAFETY: Timer operatins are locked on C side, so it is safe to operate on a
+// timer from multiple threads
+unsafe impl<T> Sync for Timer<T> {}
+
+impl<T: TimerCallback> Timer<T> {
+    /// Create an initializer for a new `Timer`
+    pub fn new() -> impl PinInit<Self> {
+        crate::pin_init!( Self {
+            timer <- Opaque::ffi_init(move |place: *mut bindings::hrtimer| {
+                // SAFETY: By design of `pin_init!`, `place` is a pointer live
+                // allication. hrtimer_init will initialize `place` and does not
+                // require `place` to be initialized prior to the call.
+                unsafe {
+                    bindings::hrtimer_init(
+                        place,
+                        bindings::CLOCK_MONOTONIC as i32,
+                        bindings::hrtimer_mode_HRTIMER_MODE_REL,
+                    );
+                }
+
+                // SAFETY: `place` is pointing to a live allocation, so the deref
+                // is safe. The `function` field might not be initialized, but
+                // `addr_of_mut` does not create a reference to the field.
+                let function: *mut Option<_> = unsafe { core::ptr::addr_of_mut!((*place).function) };
+
+                // SAFETY: `function` points to a valid allocation.
+                unsafe { core::ptr::write(function, Some(T::Receiver::run)) };
+            }),
+            _t: PhantomData,
+        })
+    }
+}
+
+#[pinned_drop]
+impl<T> PinnedDrop for Timer<T> {
+    fn drop(self: Pin<&mut Self>) {
+        // SAFETY: By struct invariant `self.timer` was initialized by
+        // `hrtimer_init` so by C API contract it is safe to call
+        // `hrtimer_cancel`.
+        unsafe {
+            bindings::hrtimer_cancel(self.timer.get());
+        }
+    }
+}
```

Comment:
If being the first field is a solution, `impl_has_timer` could become an attribute macro that also adds this field. That's not foolproof and adds proc macro complexity.

Maybe `impl_has_timer` could add a `Drop` impl for whatever type it wraps?

This kind of comes back to the `container_of` problem.

Reply:
The reason for using `Pin<&T>` to point to the target is that I have the
following arrangement:

In blk-mq, requests are handed off to the driver as a pointer to a `struct
request`. The first part of the pointee is the `struct request` and immediately
after that is a driver private area that the driver can use to store things:

```text
 +-----------------------+
 |                       |
 |                       |
 |     Request           |
 |                       |
 |                       |
 +-----------------------+
 |                       |
 |    Private            |
 |                       |
 |                       |
 +-----------------------+
```

The memory for these is pre allocated and initialized before the driver starts
accepting requests. The null block driver keeps the following in the private
area:

```rust
#[pin_data]
struct Pdu {
    #[pin]
    timer: kernel::hrtimer::Timer<Self>,
}
```

When the driver is handling a request, it will obtain a reference to this place:
`Pin<&Pdu>` that is guaranteed to be valid until the request is completed.

There are three options I think:

1. Use an owned pointer such as `Box`, `Arc`, `Aref`.
2. Somehow implement drop on the container struct, for instance with a proc macro.
3. Use a custom pointer type for this special use case - not sure how yet.


I would rather not go with 1 for my use case, because that would mean an extra
layer of indirection. I would have to store an `Arc` in my private area of the
request structure. For other uses it might be fine and the `hrtimer`
abstractions would work as they are. Just don't implement the pointer traits for
references.

For 2, we would loose the ability to eventually embed more than one timer in a
structure. I would like to keep the option to extend this patch like workqueue
to use an ID to distinguish between fields. Also, it might interfere with user
drop impls?

3 would mean unsafe code in the driver, and I am not sure how to exactly do it
yet. But maybe it could be the way.


Any input greatly appreciated :)

Reply:
> If being the first field is a solution, `impl_has_timer` could become an attribute macro that also adds this field. That's not foolproof and adds proc macro complexity.
> 
> Maybe `impl_has_timer` could add a `Drop` impl for whatever type it wraps?
> 
> This kind of comes back to the `container_of` problem.

I think that it is not sufficient, if we want to support custom drop implementations, since one could have an invariant on the struct that is violated after `drop` is called, then when dropping even the first field, the invariant is violated.

Reply:
> Any input greatly appreciated :)

I think I have an idea, but it is rather complex and needs a lot more thought:
- have a `Deconstruct` trait with `fn deconstruct(&self)` that is called at most once (done by making it `unsafe`) and guaranteed to be called before any `drop` functions from objects higher in the "contains" hierarchy are called
- call `bindings::hrtimer_cancel` inside of `Timer::deconstruct`
- somehow only allow constructions of `Timer` where the above invariant can be upheld

we might need a new smart pointer for this.

This only works if the `request` struct that you mentioned is declared on the rust side though.

Reply:
> This only works if the request struct that you mentioned is declared on the rust side though.

It's a C side structure that is initialized and destroyed from C. In this case, the private part is initialized and destroyed by calling into a vtable where Rust pointers are installed.

The following is not directly related to this PR, but it clarifies my rationale for wanting to implement the timer trait on references. I think this whole situation actually comes from me trying to achieve something that is not possible. `struct request` is actually reference counted. But when the requests are issued to the drivers, the block layer holds a reference to the request, and thus they are guaranteed to not disappear until they are completed. So for performance reasons C drivers do not take out a ref count on the structure. They just promise to only access requests that are in flight (not completed). To do this they must trust drivers to never deliver invalid completions, because at some point in the process a driver will have to turn a completion id into a request reference. If the hardware delivers an invalid completion id, you would get a reference to the wrong request and possibly have a data race.

In Rust this is a problem because we have to provide a function that turns an integer into a request reference (for getting to the request in interrupt context when hardware reports the operation done). If you provide an invalid integer (tag) here, you would get a reference that violate invariants for shared references. In Rust it is not enough to say "we trust the driver to deliver correct completion ids", because it is still possible to write safe code that has UB by simply asking for a reference for a tag that you should not be able to get.

So, I either have to make that particular function (turning an id into a request reference) unsafe, and each driver must suffer a line of unsafe code to call this, or I would have to take a reference on the request to make sure it cannot go away. The latter has the side effect of making this timer issue disappear as well, but it might hurt performance. 

In the interest of making this patch less convoluted, I think I will benchmark the cost of taking out a ref count on the request and see if that works out. Then we can skip this dance and implement the timer pointer trait for a reference counted smart pointer, probably `ARef`.

Reply:
I solved this by using an owned reference for my use case.